src/ConfigurationClientManager.ts

```diff
@@ -0,0 +1,317 @@
+        appConfigOptions?: AzureAppConfigurationOptions
+    ) {
+        let staticClient: AppConfigurationClient;
+        let options: AzureAppConfigurationOptions;
+
+        if (typeof connectionStringOrEndpoint === "string") {
+            const connectionString = connectionStringOrEndpoint;
+            options = credentialOrOptions as AzureAppConfigurationOptions;
+            this.#clientOptions = getClientOptions(options);
+            staticClient = new AppConfigurationClient(connectionString, this.#clientOptions);
+            this.#secret = parseConnectionString(connectionString, Secret);
+            this.#id = parseConnectionString(connectionString, Id);
+            // TODO: need to check if it's CDN or not
+            this.endpoint = parseConnectionString(connectionString, Endpoint);
+
+        } else if (connectionStringOrEndpoint instanceof URL) {
+            const credential = credentialOrOptions as TokenCredential;
+            options = appConfigOptions as AzureAppConfigurationOptions;
+            this.#clientOptions = getClientOptions(options);
+            staticClient = new AppConfigurationClient(connectionStringOrEndpoint.toString(), credential, this.#clientOptions);
+            this.endpoint = connectionStringOrEndpoint.toString();
+            this.#credential = credential;
+        } else {
+            throw new Error("A connection string or an endpoint with credential must be specified to create a client.");
+        }
+
+        this.#staticClients = [new ConfigurationClientWrapper(this.endpoint, staticClient)];
+        this.#validDomain = getValidDomain(this.endpoint);
+        this.isFailoverable = (options?.replicaDiscoveryEnabled ?? true) && isFailoverableEnv();
+    }
+
+    async getClients() : Promise<ConfigurationClientWrapper[]> {
+        if (!this.isFailoverable) {
+            return this.#staticClients;
+        }
+
+        const currentTime = Date.now();
+        if (this.#isFallbackClientDiscoveryDue(currentTime)) {
+            this.#lastFallbackClientRefreshAttempt = currentTime;
+            const host = new URL(this.endpoint).hostname;
+            await this.#discoverFallbackClients(host);
+        }
+
+        // Filter static clients where BackoffEndTime is less than or equal to now
+        let availableClients = this.#staticClients.filter(client => client.backoffEndTime <= currentTime);
+        // If there are dynamic clients, filter and concatenate them
+        if (this.#dynamicClients && this.#dynamicClients.length > 0) {
+            availableClients = availableClients.concat(
+                this.#dynamicClients
+                    .filter(client => client.backoffEndTime <= currentTime));
+        }
+
+        return availableClients;
+    }
+
+    async refreshClients() {
+        const currentTime = Date.now();
+        if (this.isFailoverable &&
+            currentTime > new Date(this.#lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval).getTime()) {
```

Comment:
Why don't you use `this.#isFallbackClientDiscoveryDue` here like what you did in `getClients`?

Reply:
refreshClients will be called if no availabel clients or all clients failed, these scenarios can happen when fallback isn't due.